point cloud triangulation

vcloarec opened the pull request.

vcloarec committed:
point cloud triangulation


int <--> size_t

uclaros commented:
Naive question - haven't checked any code yet: isn't there existing code in qgis that performs the Delaunay triangulation that could be reused?

vcloarec commented:
@uclaros , if you think about the `QgsDualEdgeTriangulation` class that perform a Delaunay triangulation, I am afraid the algorithm is too slow and not enough robust to support the repetitively and massive triangulation here.

There is also "poly2tri" lib already, but not sure this lib is for polygon triangulation, I don't know how this lib could support raw points without organizing them in polygons.

Maybe also GEOS ? I don't know well ... But not sure the algorithm is as fast as "delaunator". This one is very fast! And very simple!

uclaros commented on src/3d/symbols/qgspointcloud3dsymbol_p.cpp:
```diff
@@ -36,20 +36,48 @@
 #include <Qt3DRender/QEffect>
 #include <QPointSize>
 #include <QUrl>
+#include <QElapsedTimer>
```
leftover?

uclaros commented on src/3d/symbols/qgspointcloud3dsymbol_p.cpp:
```diff
@@ -36,20 +36,48 @@
 
-QgsPointCloud3DGeometry::QgsPointCloud3DGeometry( Qt3DCore::QNode *parent, unsigned int byteStride )
+#include <delaunator.hpp>
+#include <qgsproviderregistry.h>
```
leftover?

uclaros commented on src/3d/symbols/qgspointcloud3dsymbol.h:
```diff
@@ -83,8 +83,62 @@ class _3D_EXPORT QgsPointCloud3DSymbol : public QgsAbstract3DSymbol SIP_ABSTRACT
+
+    /**
+     * Returns whether triangles are filtered by size for rendering. If the triangles are filtered, triangles with a size greater
+     * than a threshold value will not be rendered, see triangleSizeThreshold().
```
I think this should clarify that the 2d triangles' *sides* are compared to the threshold value. Same for the rest of the docstrings.

uclaros commented on src/3d/shaders/pointcloud.vert:
```diff
@@ -26,6 +30,8 @@ void main(void)
     //gl_PointSize = 100.0;
 
+    worldPosition =vec3 (modelMatrix * vec4 (vertexPosition,1));
+    vertNorm=vertexNormal;
```
```suggestion
    worldPosition = vec3(modelMatrix * vec4(vertexPosition,1));
    vertNorm = vertexNormal;
```

uclaros commented on src/app/3d/qgspointcloud3dsymbolwidget.cpp:
```diff
@@ -107,6 +107,15 @@ QgsPointCloud3DSymbolWidget::QgsPointCloud3DSymbolWidget( QgsPointCloudLayer *la
+  { mTriangleSizeThresholdDoubleSpinBox->setEnabled( mFilterTriangleBySizeCheckBox->isChecked() ); } );
+  connect( mTriangleSizeThresholdDoubleSpinBox, qOverload<double>( &QDoubleSpinBox::valueChanged ), this, [&]() { emitChangedSignal(); } );
+
+  mPointSizeSpinBox->setEnabled( !mTriangulateGroupBox->isChecked() );
```
When the triangle size filter is set really low then points are rendered instead, so maybe it makes sense to leave this always enabled?

vcloarec replied:
it is a extreme case where no triangles at all are rendered, and we have the normal mode without triangulation for that. So not sure we have to keep this spin box enabled only for this extreme case, but I don't have a strong opinion on it. My point of view is that I don't like to have this spinbox that do nothing in regular case.

uclaros commented on src/ui/3d/qgspointcloud3dsymbolwidget.ui:
```diff
@@ -132,6 +132,44 @@
+           <property name="maximum">
+            <double>10000000000000000000000.000000000000000</double>
+           </property>
+          </widget>
```
A tooltip describing this is a 2d triangle's side would be nice!

uclaros commented on src/3d/symbols/qgspointcloud3dsymbol.cpp:
```diff
@@ -37,6 +37,66 @@ void QgsPointCloud3DSymbol::setPointSize( float size )
   mPointSize = size;
 }
 
+bool QgsPointCloud3DSymbol::triangulate() const
```
This reads like an action, ie *to triangulate*, maybe a name that reads as a boolean would suite better? Something like *renderAsTriangles()*?

vcloarec replied:
yes, I agree... `renderAsTriangles()` is fine.

uclaros commented on src/ui/3d/qgspointcloud3dsymbolwidget.ui:
```diff
@@ -132,6 +132,44 @@
+         <item row="0" column="0">
+          <widget class="QCheckBox" name="mFilterTriangleBySizeCheckBox">
+           <property name="text">
+            <string>Filter triangles by size</string>
```
How about *"Maximum triangle size"* instead? So that it's obvious that the biggest triangles are filtered out.

uclaros reviewed (commented):
Nice work!
Apart from some minor suggestions on the code, I've tested the feature on different datasets and had the following thoughts:
- If applied to a pointcloud with the classified renderer, the triangles that connect points of different classes are rendered black. That's probably no big deal, since the feature makes more sense to be applied to a single class and mostly ground.
- It might make sense to give the option to also filter out large vertical triangles. Not sure how that would play well with the ui though... Maybe two separate *maximum triangle size* and *maximum triangle height*?
- There's *lots* of triangles! When displaying as points the big amount of points help to create the effect of a continuous surface, but as triangles they add noise and make the scene heavy. Since in most cases we could get away with triangles from much less points than those available, would it make sense to allow skipping loading some inner node levels from the pointcloud?

vcloarec committed:
filter before triangulation


no vertical filter before and remove exterior triangles


finalize


leftover


typo, const and other leftovers


fix big triangles filter


Stefanos review

vcloarec commented:
Thanks @uclaros for this good review!

My comments:

> * If applied to a pointcloud with the classified renderer, the triangles that connect points of different classes are rendered black. That's probably no big deal, since the feature makes more sense to be applied to a single class and mostly ground

ouch! I miss this one... This is because classified renderer uses parameter value in the shader with exact color ramp. For points rendering, no problem because value is apply only on point, but with triangles, values are interpolated between vertices, so no more exact value anymore...
I am wondering if it would not be better to pass to the shader the color instead of the parameter, as for RGB. I don't know if it will be a benefit for point rendering, but for for triangles rendering, it will be . Any opinion @wonder-sk or @NEDJIMAbelgacem ? In worse case, we could have a different preparation and a RGB rendering for triangles, and keep the color ramp for classification.



> * It might make sense to give the option to also filter out large vertical triangles. Not sure how that would play well with the ui though... Maybe two separate _maximum triangle size_ and _maximum triangle height_?

We could have one, indeed... 



> * There's _lots_ of triangles!

That why some rendering are so good :smile: !


> * but as triangles they add noise and make the scene heavy

Have you some example, screen shot or dataset to share?
is is the bump effect that appears when the Z value varies a bit?

uclaros commented:
> is is the bump effect that appears when the Z value varies a bit?

Yes, I'm referring to cases where the Z precision is relatively small (larger disparity) compared to the point density, so instead of generating a floor of flat triangles you get a spiky surface like for example: 
![image](https://user-images.githubusercontent.com/11358178/155672617-90c7c3a9-e6d3-40ba-b105-685f6d5f0972.png)

Of course it's the unfiltered data to blame, but maybe we could help mitigate this while making the scene lighter at the same time.

vcloarec commented:
@uclaros 
I understand... The cons:
- if we remove some points to have a smoother surface, we will loose precision on other parts.

From my point of view, this is a data issue. We can compare that to high resolution raster DEM. Sometime the terrain is irregular like this.

vcloarec commented:
@uclaros , last commit fixes the classification rendering, if you want to have a try.

vcloarec committed:
fix point cloud triangulation with classification


sipify

vcloarec committed:
filter by height

uclaros commented:
Nice! The exact wording in the tooltips has room for improvement, maybe the native speakers can suggest something on your proper pull request!
